Added optimized ppc64le support functions for ML-KEM.

dannytsen opened the pull request:
Added optimized ppc64le support functions for ML-KEM.

The supported native functions include:
1. MLK_USE_NATIVE_NTT (ntt_ppc.S)
2. MLK_USE_NATIVE_INTT (intt_ppc.S)
3. MLK_USE_NATIVE_POLY_REDUCE (reduce.S)
4. MLK_USE_NATIVE_POLY_TOMONT (poly_tomont.S)

And other interface functions and headers.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>" .

mkannwischer committed:
Add clang_21 to compiler and constant-time tests

Signed-off-by: Matthias J. Kannwischer <matthias@kannwischer.eu>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

willieyz committed:
bitwuzla: update the bitwuzla version from 0.7.0 to 0.8.2 in nix

Signed-off-by: willieyz <willie.zhao@chelpis.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

mkannwischer committed:
Document nix 2.18 requirement; disable nix setup test for nix 2.6

Since August 2025, nixpkgs requires a nix version of at least nix 2.18.
Consequently, our nix setup tests using nix 2.6.1 and Ubuntu 22 (nix 2.6.0)
break.

This comment documents that at least nix 2.18 is required, updates
the nix test to 2.18.0, and (temporarily) disables the Ubuntu 22 test.

Signed-off-by: Matthias J. Kannwischer <matthias@kannwischer.eu>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Added optimized ppc64le support functions for ML-KEM.

The supported native functions include:
1. MLK_USE_NATIVE_NTT (ntt_ppc.S)
2. MLK_USE_NATIVE_INTT (intt_ppc.S)
3. MLK_USE_NATIVE_POLY_REDUCE (reduce.S)
4. MLK_USE_NATIVE_POLY_TOMONT (poly_tomont.S)

And other interface functions and headers.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

mkannwischer committed:
Add compiler test for zig 0.15

https://github.com/ziglang/zig/releases/tag/0.15.1

Signed-off-by: Matthias J. Kannwischer <matthias@kannwischer.eu>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker reviewed (commented):
Thank you, @dannytsen, this is an exciting contribution 🎉 

I think as the first stage of review, the goal should be to get your changes through CI, and extend it so that the PPC64 backend is exercised (to this end: do you know if your assembly works with `qemu-ppc64le`, and what flags are needed?). 

In a second phase, we can dive into the backend itself and hopefully convince ourselves that it is functionally correct and upholds the assumptions made by the frontend.

I left a few comments to kick things off, but additionally I can see that there are failures related to `autogen` and `format`, so a good starting point would be to resolve those. You should be able to run `simpasm` with a PPC cross compiler to get simplified assembly that you can check in to main source tree.

dannytsen commented on test/mk/auto.mk:
```diff
@@ -1,113 +1,33 @@
 # SPDX-License-Identifier: Apache-2.0 OR ISC OR MIT
 #
-# Automatically detect system architecture and set preprocessor flags accordingly
```
This will be my fault when I merged my code.  Will fix that.

dannytsen commented:
@hanno-becker I believe the code will work on qemu-ppc64le even though I did not run on it.  My testing platform are p9 and p10 systems.  I will go thru the comments and fix issues.  Thanks.

dannytsen committed:
Fixed auto.mk, components.mk and YML files.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker commented on mlkem/src/native/ppc64le/src/intt_ppc.S:
```diff
@@ -0,0 +1,773 @@
+
+#include "../../../common.h"
+
+.machine "any"
```
What does this do? Does it have any effect to specify "any" machine?

hanno-becker commented on dev/ppc64le/src/poly_tomont.S:
```diff
@@ -0,0 +1,163 @@
+	stxvd2x	32+7, 11, 3
+.endm
+
+.align 4
```
It seems that on PPC this means 16-byte alignment. Is this intentional, or would `.balign 4` -- 4 byte alignment -- do? The latter is what we use in all other assembly files and what `simpasm` would use by default.

hanno-becker replied:
Sorted. We use 16 byte alignment on PPC.

willieyz committed:
Add MLK_CONFIG_NO_RANDOMIZED_API with default not set

Signed-off-by: willieyz <willie.zhao@chelpis.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

willieyz committed:
Guard the `crypto_kem_keypair` and `crypto_kem_enc` from:

- `kem.c`
- `randombytes.h`
- `mlkem_native.h`

using `MLK_CONFIG_NO_RANDOMIZED_API`

Also, add a check in `common.h` to ensure `MLK_CONFIG_NO_RANDOMIZED_API` is not used together with `MLK_CONFIG_KEYGEN_PCT`

Signed-off-by: willieyz <willie.zhao@chelpis.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

willieyz committed:
Add new example `basic_deterministic`

This commit:
- Adds the `basic_deterministic` example demonstrating deterministic API usage (without a `randombytes()` implementation).
- Uses only the `crypto_kem_*_derand` functions, no `randombytes()` required.
- Updates the expected key outputs for deterministic entropy inputs (Alice: all 0 input, Bob: all 1 input).

Signed-off-by: willieyz <willie.zhao@chelpis.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

willieyz committed:
Add `--exclude example` args for "PCT enabled" CI testing

- Adds an option in the tests script to exclude specific examples.
- Needed because basic_deterministic is incompatible with MLK_CONFIG_KEYGEN_PCT.
- Allows CI to run all examples while skipping incompatible ones.

Signed-off-by: willieyz <willie.zhao@chelpis.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

willieyz committed:
CBMC: Increase the CBMC_OBJECT_BITS of `matvec_mul` to 12

- During latest change about adding derandomized config guard, the CBMC proof for `matvec_mul` failed due to SMT-solver return unknown,
  increadse the CBMC_OBJECT_BITS to fixed it.

Signed-off-by: willieyz <willie.zhao@chelpis.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

rod-chapman committed:
Introduce explicit upper bounds on lengths of
input and output buffers where appropriate.

Force CBMC to use --malloc-fail-assert for
all proofs to remove assumption on buffer lengths.

Update autogenerated files following this change.

Update Proof Guide with notes on max buffer size

Signed-off-by: Rod Chapman <rodchap@amazon.com>
Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker commented:
@dannytsen @bhess Any update on the outstanding points? (Twiddles in C, documentation of bounds)

dannytsen commented:
> @dannytsen @bhess Any update on the outstanding points? (Twiddles in C, documentation of bounds)

@hanno-becker Working on it.

hanno-becker commented:
@dannytsen @bhess Ack!

Commit:
Fixed format and styling by using autogen but no simpasm was run.
Did some more comments on value bounds in ASM files.
Used constants array instead of embedded data in assembly files.

Autogen was run under nix env on Mac.  Tests were run under HW p10.

[05:52] danny@ltcden12-lp4 mlkem_dev % ./scripts/tests func
INFO  > Functional Test    Compile     (native no_opt):  make func OPT=0 AUTO=1 -j8
INFO  > Functional Test    ML-KEM-512  (native no_opt):  make run_func_512 -j8
INFO  > Functional Test    ML-KEM-768  (native no_opt):  make run_func_768 -j8
INFO  > Functional Test    ML-KEM-1024 (native no_opt):  make run_func_1024 -j8
INFO  > Functional Test    Compile     (native opt):     make func OPT=1 AUTO=1 -j8
INFO  > Functional Test    ML-KEM-512  (native opt):     make run_func_512 -j8
INFO  > Functional Test    ML-KEM-768  (native opt):     make run_func_768 -j8
INFO  > Functional Test    ML-KEM-1024 (native opt):     make run_func_1024 -j8
All good!

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker commented on dev/ppc64le/src/intt_ppc.S:
```diff
@@ -73,6 +78,8 @@
+	#   Results word bound by 2^32 in abs value.
 	vmulosh	6, 8, V20159
 	vmulesh	5, 8, V20159
 	vmulosh	11, 12, V20159
```
What do the xxmrglw instructions do below?

hanno-becker commented on dev/ppc64le/src/intt_ppc.S:
```diff
@@ -0,0 +1,672 @@
+#__Len16:
+	#
+	# 4. len = 16, start = 0, 32, 64,,...160, 192, 224
+	#addi	14, 14, 768
```
Why is this instruction commented out?

hanno-becker commented on dev/ppc64le/src/intt_ppc.S:
```diff
@@ -0,0 +1,672 @@
+#__Len64:
+	#
+	# 6. len = 64, start = 0, 128
+	#addi	14, 14, 960
```
Same here

hanno-becker commented on dev/ppc64le/src/intt_ppc.S:
```diff
@@ -0,0 +1,672 @@
+	lxv	32+V20159, C20159_OFFSET(4)	# V20159
+	lxv	7, 0(4)	# V_25
+
+	#xxspltiw 8, 26		# for power9 and above
```
What else needs changing on power9? Which code would be commented in, which would be commented out?

hanno-becker commented on dev/ppc64le/README.md:
```diff
@@ -0,0 +1,6 @@
+
+# ppc64le backend (little endian)
+
+This directory contains a native backend for little endian POWER 8 (ppc64le) and above systems.
```
As discussed before, I couldn't yet run the code with power8 emulation, and I think you mentioned you only tested power9 and power10. Are you sure it works on power8?

hanno-becker reviewed (commented):
@dannytsen I note that you do not cache the twisted twiddles in the Montgomery multiplication: That is, in
```
        vmladduhm 15, 15, V_QINV, 3
        vmladduhm 20, 20, V_QINV, 3
        vmladduhm 25, 25, V_QINV, 3
        vmladduhm 30, 30, V_QINV, 3
```
this could all be precomputed and stored in the constant table. This is what the AArch64 and x86_64 backends do.

Have you considered that?

hanno-becker commented:
@dannytsen @bhess A general note: I don't know anything about the Power microarchitectures, but depending on how OOO they are, whether latency/throughput/units are public knowledge, and how important performance is to you, you may consider adding a microarchitecture model for [SLOTHY](https://github.com/slothy-optimizer/slothy/) and applying that here. Ordinarily, that would be a fair amount of boilerplate work, but in today's world of LLMs, I would imagine that an LLM can do this pretty quickly for you if you feed some microarchitecture documentation as input. Me or @mkannwischer can of course help with generic SLOTHY questions.

Obviously, this is not a blocker, but just an FYI for you.

dannytsen committed:
Removed unused comment lines.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker commented on mlkem/src/native/meta.h:
```diff
@@ -18,4 +18,8 @@
 #include "x86_64/meta.h"
 #endif
 
+#ifdef MLK_SYS_PPC64LE
```
@dannytsen This needs to be guarded by feature flags indicating supporting for the extensions used by the backend. Could you adjust this?

dannytsen committed:
Removed un-wanted comment.
Removed non-p8 instruction, xxspltib.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Merge branch 'pq-code-package:main' into main

hanno-becker commented:
> @hanno-becker I don't know about SLOTHY. My codes are all based my experience and understanding of the PPC microarchitecture which may not be optimal. I try to make my codes easier for me to read and maintain later but also optimize the performance. I'll spend some time on SLOTHY. Thanks.

Is there such a thing as a Software Optimization Guide for Power (e.g. like [this one](https://developer.arm.com/documentation/102160/latest/) for Cortex-A78)?

dannytsen commented:
> > @hanno-becker I don't know about SLOTHY. My codes are all based my experience and understanding of the PPC microarchitecture which may not be optimal. I try to make my codes easier for me to read and maintain later but also optimize the performance. I'll spend some time on SLOTHY. Thanks.
> 
> Is there such a thing as a Software Optimization Guide for Power (e.g. like [this one](https://developer.arm.com/documentation/102160/latest/) for Cortex-A78)?

@hanno-becker @bhess As far as I know, there are none.

hanno-becker commented on dev/ppc64le/src/intt_ppc.S:
```diff
@@ -0,0 +1,664 @@
+	MREDUCE_4X V_ZETA, V_ZETA, V_ZETA, V_ZETA, 13, 18, 23, 28
+	Write_M4C 32+13, 32+18, 32+23, 32+28
+
+.align 4
+	#
+	# Montgomery reduce loops with constant 1441
+	#
+	addi	14, 4, C1441_OFFSET
+	lvx	V1441, 0, 14
+
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 6, 7, 8, 9
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 13, 18, 23, 28
+	MWrite_8X 32+6, 32+7, 32+8, 32+9, 32+13, 32+18, 32+23, 32+28
+
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 6, 7, 8, 9
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 13, 18, 23, 28
+	MWrite_8X 32+6, 32+7, 32+8, 32+9, 32+13, 32+18, 32+23, 32+28
+
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 6, 7, 8, 9
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 13, 18, 23, 28
+	MWrite_8X 32+6, 32+7, 32+8, 32+9, 32+13, 32+18, 32+23, 32+28
+
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 6, 7, 8, 9
+	Reload_4coeffs
+	MREDUCE_4X V1441, V1441, V1441, V1441, 13, 18, 23, 28
+	MWrite_8X 32+6, 32+7, 32+8, 32+9, 32+13, 32+18, 32+23, 32+28
```
@dannytsen This will need to be moved to the beginning of the invNTT: We don't reduce the output of the base multiplication, which means that input coefficient to the invNTT is essentially unconstrained in size. Frontloading the Montgomery scaling acts as a reduction.

hanno-becker commented on integration/liboqs/ML-KEM-1024_META.yml:
```diff
@@ -9,83 +9,74 @@ length-ciphertext: 1568
 nistkat-sha256: f580d851e5fb27e6876e5e203fa18be4cdbfd49e05d48fec3d3992c8f43a13e6
 testvectors-sha256: ff1a854b9b6761a70c65ccae85246fe0596a949e72eae0866a8a2a2d4ea54b10
 principal-submitters:
-- Peter Schwabe
```
Done

hanno-becker commented on dev/ppc64le/src/intt_ppc.S:
```diff
@@ -0,0 +1,672 @@
+.align 4
+#__Len8:
+	# 3. len = 8, start = 0, 16, 32, 48,...208, 224, 240 
+	#addi	14, 14, 512
```
Ping @dannytsen can you go through all commented out instructions in the assembly files and either remove them or explain their significance?

hanno-becker commented:
@dannytsen I don't yet understand how the last two NTT layers handle data permutation. It seems to me that Layers 1-5 load and store data without interleaving. But for layer 6, the butterflies operate at distance of 4 coefficients, which is smaller than the vector size (8 coefficients), so I would expect some intra-vector rearrangement or a modified load (like ld2 on Arm). I see that the _store_ operations do some interleaving, but I don't see anything up until and _including_ the load for layer 5. What am I missing?

dannytsen commented:
> @dannytsen I don't yet understand how the last two NTT layers handle data permutation. It seems to me that Layers 1-5 load and store data without interleaving. But for layer 6, the butterflies operate at distance of 4 coefficients, which is smaller than the vector size (8 coefficients), so I would expect some intra-vector rearrangement or a modified load (like ld2 on Arm). I see that the _store_ operations do some interleaving, but I don't see anything up until and _including_ the load for layer 6. What am I missing?

@hanno-becker I did not read x86 or arm codes.  I just coded it based on the C-file.  There maybe something I missed and can be improved.  I'm not mathematician so there will be some mistakes or incorrect approach.

Layer 6 could be re-worked.  The data was re-arranged in the beginning and re-arrange back which may be more work needs to be done.  I definitely will look more into it.  Thanks.

dannytsen commented:
@hanno-becker Thanks for taking so much on ppc64le integration.  I do learn a lot from here in all aspects.   I don't know how much time will take me but I'll work on it.

hanno-becker commented:
> @hanno-becker Thanks for taking so much on ppc64le integration. I do learn a lot from here in all aspects. I don't know how much time will take me but I'll work on it.

@dannytsen This is great to hear! Please don't hesitate to ask if anything is unclear about the algorithmic aspects of the AArch64 implementation, or how to translate it. Otherwise, I'm looking forward to seeing the updated code!

dannytsen commented:
@hanno-becker Still working on fixing current code before the other.

dannytsen committed:
Fixed failure in INTT unit tests.
Re-arranged zeta array for NTT/INTT for Len 2 and 4.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Merge branch 'pq-code-package:main' into main

dannytsen committed:
Merge branch 'main' into main

dannytsen committed:
Fixed illegal instructions used for p8 arch, lxv and stxv.
Also fixed instruction byte orerding mismatch for p8 and p9/10, lxvx/stxvx
and lxvd2x/stxvd2x.  Used lxvd2x and stxvd2x for consistant byte ordering.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Merge branch 'pq-code-package:main' into main

hanno-becker commented on dev/ppc64le/src/ntt_ppc.S:
```diff
@@ -0,0 +1,515 @@
+	addi	18, 16, \next
+	addi	19, 17, \step
+	addi	20, 18, \next
+	addi	21, 19, \step
```
Half of these offsets are not used here, but further down in Loda_4AJ, which is confusing. The code should either be moved, or a comment added.

hanno-becker reviewed (changes requested):
@dannytsen Thank you again for your work.

I continued reviewing, but quickly found myself stuck without extensive help from an LLM commenting and rewriting the code for me. As it stands, the code is too difficult to understand for @mkannwischer and myself to commit to maintaining it.

The following steps would greatly help to make the code more understandable and maintainable:
* Addition of `#define ...` abbreviations for all the registers in use. Following the raw numeric register identifiers is difficult, as the reader does not (yet) have the mental model that you had when writing the code. This is exacerbated by the fact that the there is no way to tell (without knowledge of the instruction signature) whether a numeric value is a register identifier or an immediate. For example, in `NTT_MREDUCE_4X 5, 16, 16, ...`, `5` is a register identifier but `16` is an immediate. 
* More extensive comments, and fixing of existing comments. There are a number of comments which give a sense of what happens, but are not precise enough to _exactly_ understand what's happening; see the review comments.

Finally, can you also switch to using `/* ... */` style comments rather than `# ...`, matching other assembly files in mlkem-native? Also, could you untabify the source (use spaces instead of tabs exclusively)?

dannytsen committed:
Merge branch 'main' into main

Signed-off-by: Danny Tsen <88787197+dannytsen@users.noreply.github.com>

dannytsen commented:
> @dannytsen Thank you again for your work.
> 
> I continued reviewing, but quickly found myself stuck without extensive help from an LLM commenting and rewriting the code for me. As it stands, the code is too difficult to understand for @mkannwischer and myself to commit to maintaining it.
> 
> The following steps would greatly help to make the code more understandable and maintainable:
> 
>     * Addition of `#define ...` abbreviations for all the registers in use. Following the raw numeric register identifiers is difficult, as the reader does not (yet) have the mental model that you had when writing the code. This is exacerbated by the fact that the there is no way to tell (without knowledge of the instruction signature) whether a numeric value is a register identifier or an immediate. For example, in `NTT_MREDUCE_4X 5, 16, 16, ...`, `5` is a register identifier but `16` is an immediate.
> 
>     * More extensive comments, and fixing of existing comments. There are a number of comments which give a sense of what happens, but are not precise enough to _exactly_ understand what's happening; see the review comments.
> 
> 
> Finally, can you also switch to using `/* ... */` style comments rather than `# ...`, matching other assembly files in mlkem-native? Also, could you untabify the source (use spaces instead of tabs exclusively)?

@hanno-becker Sure.  I'll fix as much as I can.

dannytsen committed:
Merge branch 'pq-code-package:main' into main

dannytsen committed:
This commit fixed comments and added more comments and others,

1. De-tabified.
2. Merged next and step used in macro.
3. Used immediate values for offsets used in macro.
4. More comments explaining the operation and contents.
5. Changed the comment style.

In this commit, numeric register identifiers have not been fixed yet.
will do that next.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Merge branch 'pq-code-package:main' into main

dannytsen committed:
Merge branch 'main' into main

hanno-becker commented:
How are you getting on @dannytsen?

dannytsen committed:
Merge branch 'main' into main

dannytsen committed:
Addition of #define for all registers used.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Merge branch 'main' into main

hanno-becker commented:
@dannytsen Apologies if I was not clear, but the ask was not to have prefix names `v0,v1,...` instead of the numerical ones, but to have names which express the meaning of a variable. For example, `data0, data1, ...` for data vectors, or `twiddle`/`root` or similar for multiplication constants. Please have a look into the existing 'clean' assembly files, e.g. https://github.com/pq-code-package/mlkem-native/blob/main/dev/aarch64_clean/src/ntt.S.

dannytsen commented:
> How are you getting on @dannytsen?

@hanno-becker Updated.

> @dannytsen Apologies if I was not clear, but the ask was not to have prefix names `v0,v1,...` instead of the numerical ones, but to have names which express the meaning of a variable. For example, `data0, data1, ...` for data vectors, or `twiddle`/`root` or similar for multiplication constants. Please have a look into the existing 'clean' assembly files, e.g. https://github.com/pq-code-package/mlkem-native/blob/main/dev/aarch64_clean/src/ntt.S.

@hanno-becker I'll take a look when I have some cycle.  Thanks.

dannytsen committed:
This patch fixes the following,

1. Added detailed comments on NTT and INTT implementations.
2. Used C type symbols to improve readability.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker commented:
Thanks a lot @dannytsen for your work, this is looking good!

Could you extend your use of aliases to also cover parts like
```asm
.macro Load_4Rjp
        lxvd2x  32+vdata_b1, 3, 10     /* V8: vector r'0 */
        lxvd2x  32+vdata_b2, 3, 17    /* V12: vector for r'1 */
        lxvd2x  32+vdata_b3, 3, 19    /* V16: vector for r'2 */
        lxvd2x  32+vdata_b4, 3, 21    /* V20: vector for r'3 */

        lxvd2x  32+vdata_a1, 3, 9     /* V21: vector r0 */
        lxvd2x  32+vdata_a2, 3, 16    /* V22: vector r1 */
        lxvd2x  32+vdata_a3, 3, 18    /* V23: vector r2 */
        lxvd2x  32+vdata_a4, 3, 20    /* V24: vector r3 */
.endm
```

We should have as little hardcoded registers as possible.

dannytsen committed:
Used macro and removed ntt loops for len=2 and 4.
Replaced more resgiters number with C type variable names.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

dannytsen committed:
Fixed copyright header.

Signed-off-by: Danny Tsen <dtsen@us.ibm.com>

hanno-becker commented on dev/ppc64le/src/ntt_ppc.S:
```diff
@@ -0,0 +1,408 @@
+	vsrah 13, 15, 4		# >> 1
+	vsrah 18, 20, 4		# >> 1
+	vsrah 23, 25, 4		# >> 1
+	vsrah 28, 30, 4		# >> 1
```
@dannytsen Could you comment on the origin of this code? Were you following a particular source? As mentioned above, the algorithm is originally described in https://eprint.iacr.org/2021/986 but therein unnecessarily constrained to at least one operand being odd. Were you aware that you were relaxing this constraint here? Any recollections on how this code came about are appreciated.

dannytsen commented:
This is derived from the C code which is commented on the top of the macro.  Just use the Power instruction to accomplish >> 16.  Right shift 15 bits and then 1 bit.

 * montgomery_reduce
 * t = a * QINV
 * t = (a - (int32_t)t*_MLKEM_Q) >> 16

Vmhraddshs instruction right shifts 15 bits so we need to right shift one more bit.

The result of the instruction is,
r e s u l t ← ( ( ( s r c 1  s r c 2 ) + 0 x 0000_4000 ) >> 15 ) + s r c 3

Vsrah then right shift one bit

________________________________
From: Hanno Becker ***@***.***>
Sent: Sunday, May 17, 2026 11:11 PM
To: pq-code-package/mlkem-native ***@***.***>
Cc: Danny Tsen ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [pq-code-package/mlkem-native] Added optimized ppc64le support functions for ML-KEM. (PR #1184)

@ hanno-becker commented on this pull request. In dev/ppc64le/src/ntt_ppc. S: > + # fqmul = zeta * coefficient + # Modular multification bond by 2^16 * q in abs value + vmladduhm 15, 13, \_vz0, 3 + vmladduhm 20, 18, \_vz1, 3 + vmladduhm 25,


@hanno-becker commented on this pull request.

________________________________

In dev/ppc64le/src/ntt_ppc.S<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_pq-2Dcode-2Dpackage_mlkem-2Dnative_pull_1184-23discussion-5Fr3256296096&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=0yWke21PNfVf_fP-IAvlSCVU6_ftL79eJhKTK3XKWho&e=>:

> +     # fqmul = zeta * coefficient
+       # Modular multification bond by 2^16 * q in abs value
+       vmladduhm 15, 13, \_vz0, 3
+       vmladduhm 20, 18, \_vz1, 3
+       vmladduhm 25, 23, \_vz2, 3
+       vmladduhm 30, 28, \_vz3, 3
+
+       # Signed multiply-high-round; outputs are bound by 2^15 * q in abs value
+       vmhraddshs 14, 13, \_vz0, 3
+       vmhraddshs 19, 18, \_vz1, 3
+       vmhraddshs 24, 23, \_vz2, 3
+       vmhraddshs 29, 28, \_vz3, 3
+
+       vmladduhm 15, 15, V_QINV, 3
+       vmladduhm 20, 20, V_QINV, 3
+       vmladduhm 25, 25, V_QINV, 3
+       vmladduhm 30, 30, V_QINV, 3
+
+       vmhraddshs 15, 15, V_NMKQ, 14
+       vmhraddshs 20, 20, V_NMKQ, 19
+       vmhraddshs 25, 25, V_NMKQ, 24
+       vmhraddshs 30, 30, V_NMKQ, 29
+
+       vsrah 13, 15, 4         # >> 1
+       vsrah 18, 20, 4         # >> 1
+       vsrah 23, 25, 4         # >> 1
+       vsrah 28, 30, 4         # >> 1


@dannytsen<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dannytsen&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=v5ME1benQJiru4wanRSa_kcQfLA78nTY1dfAVj6ldVQ&e=> Could you comment on the origin of this code? Were you following a particular source? As mentioned above, the algorithm is originally described in https://eprint.iacr.org/2021/986<https://urldefense.proofpoint.com/v2/url?u=https-3A__eprint.iacr.org_2021_986&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=zfUr7Cr7pMwcjomt0ImD7-EW2biPHxdjF17hSVQXUOk&e=> but therein unnecessarily constrained to at least one operand being odd. Were you aware that you were relaxing this constraint here? Any recollections on how this code came about are appreciated.

—
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_pq-2Dcode-2Dpackage_mlkem-2Dnative_pull_1184-23discussion-5Fr3256296096&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=0yWke21PNfVf_fP-IAvlSCVU6_ftL79eJhKTK3XKWho&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AVFMR7NFKK4IAVLEWJRROV343KEN7AVCNFSM6AAAAACGBE7QN2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHM2DGMBXGE4DKNRVGY&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=HIKo_IIvAcxdwYqI8TOskiKPPA81YIlxSdZNqGQZ5D0&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=iCuaFDrachzw62XJ848BTzbkX6GrY_VlJK9P8FDdBtQ&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=BSDicqBQBDjDI9RkVyTcHQ&r=zspFcGYEyUrRywX_TdjlLwwrCx0eBFnzcs6XZVVVMh0&m=boOpR1VW8Lxtplz19tFpOX7rpgJxsQs_1oHpj46AxOj9j2hmPpwgVMaryVRtuq1N&s=bJArcLdwxeGymAc4ApxJ2a5XCR0nS9-SGSKYnNOv0xk&e=>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

hanno-becker commented:
@dannytsen The relation between the C and the ASM is not at all obvious -- the multiplication instructions are doubling high multiplies with rounding, whereas the C code is a plain high multiplication. The correction from two different rounding operations are used to emulate the correction needed from what's actually happening, namely a 32-bit addition where the low 16-bits can overflow into the upper 32-bit. If the compiler really turned this into the rounding Montgomery by itself, without this even being documented anywhere in the literature, this would be pretty amazing. Which compiler is that?

dannytsen commented:
@hanno-becker the instruction is documented in the ISA 2.07 and above.

hanno-becker commented:
Yes, the _instruction_ is documented, but it is entirely unclear that it would apply to the C code that's being assembled. The C code is a widening 16x16->32 bit multiply-and-add, while what's being used on the ASM here are two rounding+doubling high multiplications, with the rounding taking care of the carry-in from low to high bits in the 32-bit add.

cc @bhess This is stunning. Can you explain this?

bhess commented:
I'm not certain about the origin of the code, but it appears correct to me.

Per Power ISA, we have

- `vmladduhm(a,b,c) = (ab + c) mod 2^16`
- `vmhraddshs(a,b,c) = saturate16( floor((ab + 2^14)/2^15) + c )` — Vector Multiply-High-Round-Add Signed Halfword Saturate
- `vsrah(x, 1) = floor(x/2)` — arithmetic right shift, rounds toward -Inf (floor)

The four steps are:

| step | instr | result |
|---|---|---|
| A | `vmladduhm tmp, b, z, vr3` | $\mathrm{lo}(bz)$ |
| B | `vmhraddshs H, b, z, vr3` | $H = \lfloor (bz + 2^{14})/2^{15}\rfloor$ |
| C | `vmladduhm t, tmp, V_QINV, vr3` | $t$ |
| D | `vmhraddshs R, t, V_NMKQ, H` | $R = \lfloor(-tq + 2^{14})/2^{15}\rfloor + H$ |
| E | `vsrah n, R, 1` | $\lfloor R/2\rfloor$ |

Here $t: = (bz \cdot q^{-1}) \bmod 2^{16}$ is the signed Montgomery quotient, so by construction $bz - tq = 2^{16}\cdot n$ with $n = \mathrm{mont}(bz)$.

In B we can write
$H \=\ \left\lfloor\tfrac{bz+2^{14}}{2^{15}}\right\rfloor \=\ \left\lfloor\tfrac{bz-2^{14}}{2^{15}}\right\rfloor + 1.$

We have $-tq = 2^{16}n - bz = 2^{15}\cdot 2n - bz$. Substitute that into the floor in Step D:

$\left\lfloor\tfrac{-tq + 2^{14}}{2^{15}}\right\rfloor\=\ \left\lfloor 2n - \tfrac{bz - 2^{14}}{2^{15}}\right\rfloor\=\ 2n - \left\lceil\tfrac{bz-2^{14}}{2^{15}}\right\rceil .$

Using the two steps above we can write

$R \=\ 2n \+\ \underbrace{\left\lfloor\tfrac{bz-2^{14}}{2^{15}}\right\rfloor + 1 - \left\lceil\tfrac{bz-2^{14}}{2^{15}}\right\rceil}_{\delta \,\in\, \{0,1\}}.$

$\delta = 1$ iff $bz - 2^{14}$ is divisible by $2^{15}$ (floor = ceiling), and $0$ otherwise (ceiling exceeds floor by exactly 1).
Therefore step E gives $\lfloor R/2 \rfloor = \lfloor (2n+\delta)/2\rfloor = n$ for both $\delta \in \{0,1\}$, including negative $n$, this is why `vsrah` (floor) is required.

Finally, the saturation in `vmhraddshs` is not reached since the Kyber input ranges are inside int16 limits.

hanno-becker commented:
@bhess I agree the code is correct -- I am just stunned that the compiler does this reasoning and figures out that two rounding-multiply operations can be used to emulate a plain multiply-add. As I said, we did document the approach in the Neon NTT paper, but erroneously said one operand would need to be odd. 

So what I asked "Can you explain this?" I was wondering if you have any idea how/why the PPC compiler is able to do this.

bhess commented:
> So what I asked "Can you explain this?" I was wondering if you have any idea how/why the PPC compiler is able to do this.

Yeah, this would be quite surprising if a compiler were able to do this. I asked around and indeed, the colleagues derived the method independently for the POWER implementation, without knowing your paper, so no compiler magic involved.

hanno-becker commented:
> I asked around and indeed, the colleagues derived the method independently for the POWER implementation

Brilliant, thank you for finding that out. Well done to the team, then :+1:
